### ISSUE TITLE
Unsigned transaction serialization not exposed for EIP-155 use

### ISSUE THREAD

**Issue by alexfmpe** [ethereumjs/ethereumjs-monorepo#1188]
When signing with a ledger app that parses the transaction itself, it is necessary to produce a serialization of a EIP-155 unsigned transaction. `_getMessageToSign` computes that serialization, but only returns the hash
https://github.com/ethereumjs/ethereumjs-monorepo/blob/00fc597a817edda5dfb0673027ead3a21dd309d4/packages/tx/src/legacyTransaction.ts#L137-L154

whereas `raw` always treats the transaction as signed 
https://github.com/ethereumjs/ethereumjs-monorepo/blob/00fc597a817edda5dfb0673027ead3a21dd309d4/packages/tx/src/legacyTransaction.ts#L112-L124

The current workaround is to reimplement `_getMessageToSign` downstream but without the hashing.

**Comment by holgerd77** [comment]
Thanks for reporting!

For reference: [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/blockchain-reorg-support/packages/tx/src/transaction.ts) is the implementation of the legacy tx before the EIP-2930/Typed Tx refactor.

We might want to reintroduce a public `getMessageToSign()` method (respectively change the return type + name of the `_getMessageToSign()` method?). Not sure though on first sight, do not have the complete overview yet.

Also, as some reference: [EIP-155](https://eips.ethereum.org/EIPS/eip-155)

This is something which we should address relatively soon.

**Comment by holgerd77** [comment]
> We might want to reintroduce a public `getMessageToSign()` method (respectively change the return type + name of the `_getMessageToSign()` method?). Not sure though on first sight, do not have the complete overview yet.

Ok, this was not a useful suggestion on second look.

@alexfmpe just to confirm: are you updating from an old version of the library (and if so: which version?) or is this a "new" use case you are trying to solve?

Also for confirmation, now targeting anyone:  Use case of the `raw` method covered with `v`, `r` and `s` being `undefined` is for serialization of an unsigned tx, which is e.g. used in some RPC methods I guess? Is this correct?

Would it make sense to add some parameter `raw=false` to `getMessageToSign()` which would return the raw un-hashed format if set to `true`? (not sure, did we actually already had something like this in some old version of the library? 😛)

//cc @cgewecke @ryanio @jochem-brouwer @alcuadrado

**Comment by alexfmpe** [comment]
> @alexfmpe just to confirm: are you updating from an old version of the library (and if so: which version?) or is this a "new" use case you are trying to solve?

I am and have been using version 3.0.0, which has this issue. I've ran into it a second time and made an issue. Would updating affect it?

**Comment by holgerd77** [comment]
I am not understanding the question.

**Comment by alexfmpe** [comment]
I am not updating from an old version (I'm still on the now-old 3.0.0 version) and was trying to understand why that mattered.

**Comment by holgerd77** [comment]
@ryanio can you have a look here (others as well)? Might it be the case that our behavior for the `raw()` method (which was also already there in the old-but-refactored implementation linked above) for unsigned txs is just wrong and not taking `EIP-155` into account?

Not sure, I always had my problems with the concrete definition of the `raw()` method, but isn't this the data format which prepares for a serialization (if we have a final answer on this we should likely add to the code docs)? And in that case the format would be wrong, actually for both unsigned cases, including `EIP-155` (this should use the values defined in the EIP) and excluding (this should just serve 6 instead of 9 elements).

Hmm, not sure. Also looked for the usage of the `raw()` method in the monorepo, seems we are just not using this for unsigned cases.

//cc @alcuadrado

**Comment by alcuadrado** [comment]
Hey @alexfmpe! 

We definitely missed this when designing the new version of the Transaction class. It makes total sense for Ledger to require the preimage of the hash, as otherwise users would be signing possibly dangerous data and they wouldn't be able to show the confirmation dialog.

I think the best way to solve this would be to change the signature of the method from `getMessageToSign()`  to `getMessageToSign(hashMessage = true)`.

**Comment by alcuadrado** [comment]
> Would it make sense to add some parameter `raw=false` to `getMessageToSign()` which would return the raw un-hashed format if set to `true`?

oh, now i get this. Yes, it's pretty much the same, but with a different name which better describes what it does.

> Might it be the case that our behavior for the `raw()` method (which was also already there in the old-but-refactored implementation linked above) for unsigned txs is just wrong and not taking `EIP-155` into account?

I think raw is "fine". It's name is not really clear, but it returns the array of RLP encoded values (i.e. `rlp.decode(tx.serialize())`).  This is what you hash and sign before EIP-155. EIP-155 explicitly stops doing that, separating the concept of data to hash from the rlp serialization of the tx.

**Comment by brad-decker** [comment]
Hi there, just sharing my experience with the same issue.

to solve in userland I create an unfrozen version of the transaction:
```js
  const unfrozenTx = TransactionFactory.fromTxData(tx.toJSON(), { common: tx.common, freeze: false })
  unfrozenTx.v = new ethUtil.BN(ethUtil.addHexPrefix(tx.common.chainId()), 'hex')
```

I can then `serialize()` the `unfrozenTx` and send that to ledger for signing. This results in an EIP-155 compliant `v` value returned.

I then construct another new transaction, restoring the original frozen status of the transaction before passing it back to the caller

```js
 const txData = tx.toJSON()
 // The fromTxData utility expects v,r and s to be hex prefixed
 txData.v = ethUtil.addHexPrefix(payload.v)
 txData.r = ethUtil.addHexPrefix(payload.r)
 txData.s = ethUtil.addHexPrefix(payload.s)
 // Adopt the 'common' option from the original transaction and set the
 // returned object to be frozen if the original is frozen.
 return TransactionFactory.fromTxData(txData, { common: tx.common, freeze: Object.isFrozen(tx) })
```

**Comment by holgerd77** [comment]
Here is a proposed fix: #1214

Open for review.

**Comment by holgerd77** [comment]
We have just released this as tx `v3.1.4` in #1217, it would be great if you can give a short feedback here if this is working for you.
